### ISSUE TITLE
feat: Native multi-label classification support via `multi_label` flag (sigmoid + BCE)

### ISSUE THREAD

**Issue by surajsharan** [ultralytics/ultralytics#24269]
### Search before asking

- [x] I have searched the Ultralytics [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

> Note: This replaces #24268 which was accidentally closed as duplicate by the author.
### Search before asking
- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no active implementation.

---

### Is your feature request related to a problem?

Multi-label image classification — where a single image can belong to multiple categories simultaneously — is not natively supported in the Ultralytics classification pipeline. The current `task=classify` uses a softmax head + cross-entropy loss, which assumes mutually exclusive classes.

This has been a recurring community request spanning **2+ years across multiple YOLO generations**:

- #489 — YOLOv8 multilabel train/inference (Jan 2023, closed as stale)
- #8529 — Multilabel classification possibility (Feb 2024, closed as stale)  
- #17829 — YOLOv11 multilabel support (Nov 2024, closed unanswered)
- #22478 — YOLOv11/12 multilabel training (Oct 2025, closed as stale)
- #16015 — Only implementation attempt (Sep 2024, closed when author deleted their account — never merged)

None of these were resolved with a working, merged implementation.

---

### Describe the solution you'd like

A `multi_label` boolean flag in the training configuration that switches the classify pipeline from single-label to multi-label mode — **with no breaking changes to existing behavior**.

**Core changes (flag-based, minimal surface area):**

| Component | Single-label (default) | Multi-label (`multi_label=True`) |
|---|---|---|
| Head activation | Softmax | Sigmoid |
| Loss | CrossEntropyLoss | BCEWithLogitsLoss |
| Target format | Class index (int) | Multi-hot vector (float) |
| Metrics | Top-1/Top-5 accuracy | Per-class AP + mAP |
| Inference output | Single argmax class | All classes above threshold |

**Usage (proposed API):**
```python
# Training
model = YOLO("yolo11n-cls.pt")
model.train(data="multilabel_dataset.yaml", multi_label=True, epochs=100)

# Inference
results = model.predict("image.jpg")  # returns per-class sigmoid probabilities
```

**Dataset format (YAML + labels CSV):**
```yaml
# multilabel_dataset.yaml
path: ./dataset
train: images/train
val: images/val
nc: 5
names: ['cat', 'dog', 'outdoor', 'indoor', 'animal']
```

### Use case

Multi-label classification is a fundamental need in real-world CV pipelines where objects/scenes belong to multiple categories simultaneously. Common scenarios include:

- **Retail/warehouse**: A shelf image tagged as both "electronics" + "damaged" + "out-of-stock"
- **Medical imaging**: A scan flagged for multiple conditions at once
- **Content moderation**: Images with multiple attribute tags (e.g. "outdoor", "crowded", "nighttime")
- **E-commerce**: Product images with multiple category tags

Currently users hitting this need either (a) fork and patch manually, (b) use a separate framework entirely, or (c) ask in issues and get no resolution — as seen across #489, #8529, #17829, #22478 spanning 2+ years. This feature would serve a large segment of the classification user base that deals with non-mutually-exclusive label spaces.

### Additional

**Design rationale vs PR #16015:**

The previous attempt (#16015) introduced entirely separate classes — `MultiLabelClassify` head, `MultiLabelClassificationTrainer/Validator/Predictor` — plus a new `torcheval` dependency. High surface area, low maintainability, never merged.

This proposal is the opposite:
- No new head class — reuses existing `Classify` head, sigmoid activated via flag
- No new trainer/validator/predictor — `if multi_label` branches in existing classify classes
- No new external dependencies — uses native `BCEWithLogitsLoss`
- Fully backward compatible — `multi_label=False` (default) preserves all existing behavior
- Explicitly fixes the `plotting.py` `cls[idx]` dimension mismatch bug from #16015

**I have a working implementation ready.** Submitting the PR immediately after this issue. The implementation has been tested end-to-end with a multilabel dataset.

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @surajsharan, thank you for your interest in Ultralytics 🚀 and for the detailed feature request around native multi-label classification support in YOLO11 classify. This is an automated response to help get things moving quickly, and an Ultralytics engineer will also assist you soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

Since this is a feature request ✨, thank you as well for outlining the proposed API, design considerations, related issues, and your willingness to submit a PR. If you do open the PR, please include a clear summary of the proposed behavior, affected classification components, example dataset formatting, validation results, and any tests or benchmarks that help reviewers evaluate the change efficiently.

If this later turns out to involve a 🐛 Bug Report in current classification behavior, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by surajsharan** [comment]
@glenn-jocher 

Working implementation is ready. Submitting PR shortly.
Happy to address any design feedback

**Comment by glenn-jocher** [comment]
Sounds good — please open the PR with the change kept narrowly scoped to the existing `classify` pipeline, avoid broad `default.yaml` surface changes if possible, and include focused tests plus any needed docs updates per the [contributing guide](https://docs.ultralytics.com/help/contributing/).

**Comment by surajsharan** [comment]
> Sounds good — please open the PR with the change kept narrowly scoped to the existing `classify` pipeline, avoid broad `default.yaml` surface changes if possible, and include focused tests plus any needed docs updates per the [contributing guide](https://docs.ultralytics.com/help/contributing/).

Thanks @glenn-jocher  PR is already open at #24270.

The implementation is narrowly scoped:
- No new head classes, trainer, validator, or predictor classes
- `if multi_label` branches inside existing classify pipeline only
- `multi_label: False` added to `default.yaml` under the `# Classification` 
  section, following the same pattern as `dropout: 0.0` (classify-only 
  setting, no classify-specific config file exists)

Docs updated in docs/en/tasks/classify.md and 
  docs/en/datasets/classify/index.md covering dataset format, 
  train/val/predict examples (Python + CLI), and metrics

Happy to address any further review comments.

**Comment by glenn-jocher** [comment]
Thanks for opening `#24270` — continuing on the PR is

**Comment by sarpx** [comment]
I'm eagerly awaiting this feature.

**Comment by manan45** [comment]
this was missing in repo🚀🚀

**Comment by glenn-jocher** [comment]
Thanks for the interest — the implementation is now in [PR #24270](https://github.com/ultralytics/ultralytics/pull/24270), so please follow that thread for review updates and any testing feedback.

**Comment by surajsharan** [comment]
PR #24270 is open and actively being reviewed , keeping this issue open 
for reference until the PR merges.

**Comment by glenn-jocher** [comment]
Thanks, that works — please keep implementation details, test results, and follow-up in `#24270` so review stays centralized, and we can leave this issue as the feature request reference until the PR is resolved.
